File: java/src/main/java/HAReads.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.sql.SQLException;

public class HAReads {
    private static final String URL =
        "jdbc:postgresql://localhost:8812,localhost:8813,localhost:8814/qdb"
        + "?targetServerType=any&connectTimeout=3";

    private static final String USER = "admin";
    private static final String PASS = "quest";

    private Connection conn;  // persistent connection reference

    private Connection getConnectionWithRetry() {
        while (true) {
            try {
                Connection c = DriverManager.getConnection(URL, USER, PASS);
                c.setAutoCommit(true);
                System.out.println("Connected to " + c.getMetaData().getURL());
                return c;
            } catch (SQLException e) {
                System.err.println("Connection attempt failed: " + e.getMessage());
                try {
                    Thread.sleep(2000); // wait 2s before retry
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException("Interrupted while waiting to retry", ie);
                }
            }
        }
    }

    public String runQuery(String sql) throws SQLException {
        if (conn == null || conn.isClosed()) {
            conn = getConnectionWithRetry();
        }
        try (Statement st = conn.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            if (rs.next()) {
                return rs.getString(1);
            }
            return null;
        } catch (SQLException e) {
            // on error, reconnect with retry loop and retry query
            System.err.println("Connection error, retrying: " + e.getMessage());
            if (conn != null) {
                try { conn.close(); } catch (SQLException ignore) {}
            }
            conn = getConnectionWithRetry();
            try (Statement st = conn.createStatement();
                 ResultSet rs = st.executeQuery(sql)) {
                if (rs.next()) {
                    return rs.getString(1);
                }
                return null;
            }
        }
    }

    public void close() {
        if (conn != null) {
            try { conn.close(); } catch (SQLException ignore) {}
            conn = null;
        }
    }

    public static void main(String[] args) throws Exception {
        HAReads demo = new HAReads();
        String query = "select value from (show parameters) "
                     + "where property_path = 'cairo.wal.temp.pending.rename.table.prefix';";

        for (int i = 1; i <= 250; i++) {
            String value = demo.runQuery(query);
            System.out.printf("Query %3d -> %s%n", i, value);
            Thread.sleep(300); // 300ms between queries
        }

        demo.close();
    }
}

File: python/ha_reads.py
# ha_reads.py
import time
import psycopg
from psycopg import OperationalError, InterfaceError

class HAReads:
    CONN_STR = (
        "postgresql://admin:quest@"
        "localhost:8812,localhost:8813,localhost:8814/qdb"
        "?target_session_attrs=any&connect_timeout=3"
    )

    def __init__(self):
        self.conn: psycopg.Connection | None = None

    def get_connection_with_retry(self) -> psycopg.Connection:
        """Keep trying until a connection succeeds."""
        while True:
            try:
                conn = psycopg.connect(self.CONN_STR, autocommit=True)
                print(f"Connected to {conn.info.host}:{conn.info.port}")
                return conn
            except OperationalError as e:
                print(f"Connection attempt failed: {e}")
                time.sleep(2)

    def run_query(self, sql: str, params=None):
        """Run query, reconnect on connection error, retry once."""
        if self.conn is None or self.conn.closed:
            self.conn = self.get_connection_with_retry()

        try:
            with self.conn.cursor() as cur:
                cur.execute(sql, params)
                return cur.fetchall()
        except (OperationalError, InterfaceError) as e:
            print(f"Connection error, retrying: {e}")
            try:
                if self.conn and not self.conn.closed:
                    self.conn.close()
            except Exception:
                pass
            self.conn = self.get_connection_with_retry()
            with self.conn.cursor() as cur:
                cur.execute(sql, params)
                return cur.fetchall()

    def close(self):
        if self.conn and not self.conn.closed:
            try:
                self.conn.close()
            except Exception:
                pass
            self.conn = None

if __name__ == "__main__":
    demo = HAReads()
    query = (
        "select value from (show parameters) "
        "where property_path IN ( 'replication.role', 'cairo.wal.temp.pending.rename.table.prefix') limit 1;"
    )

    for i in range(1, 251):  # 250 iterations
        rows = demo.run_query(query)
        value = rows[0][0] if rows else None
        print(f"Query {i:3d} -> {value}")
        time.sleep(0.3)  # 300 ms

    demo.close()
